File: sample/sample-java/SampleAnalyzeTS.java
//---------------------------------------------------------------------------
//
// TSDuck sample Java application : analyze a transport stream.
// A TSProcessor is used, using plugin "analyze" and JSON output.
// The JSON output is intercepted and parsed in a Java class.
// When found, the list of services in the transport stream is displayed.
// Any other information on the TS can be similarly extracted and used.
//
// Warning: need org.json third-party package.
// Source: https://github.com/stleary/JSON-java
// Windows: execute get-org-json.ps1
// Unix (Linux, macOS): execute get-org-json.sh
//
//----------------------------------------------------------------------------

import org.json.JSONArray;
import org.json.JSONObject;

import io.tsduck.AbstractAsyncReport;
import io.tsduck.Report;
import io.tsduck.TSProcessor;

public class SampleAnalyzeTS {

    /**
     * This string is a user-defined marker to locate the JSON line in the log.
     * It can be anything that is sufficiently weird to be unique in the logs.
     */
    private static final String JSON_MARKER = "[_TS_JSON_]";

    /**
     * This method processes the parsed JSON data from the TS analysis.
     * Here, we just display the list of services.
     * @param jsonText TS analysis in a JSON string.
     */
    private static void jsonHandler(String jsonText) {
        final JSONObject root = new JSONObject(jsonText);
        final JSONArray services = root.getJSONArray("services");
        for (int i = 0; i < services.length(); i++) {
            JSONObject srv = (JSONObject) services.get(i);
            System.out.printf("Name: \"%s\", provider: \"%s\", bitrate: %d b/s\n", srv.getString("name"), srv.getString("provider"), srv.getInt("bitrate"));
        }
    }

    /**
     * A pure Java class which handles TSDuck log messages.
     */
    private static class Logger extends AbstractAsyncReport {

        /**
         * Constructor.
         * @param severity Initial severity.
         */
        public Logger() {
            super(Report.Info, false, 512);
        }

        /**
         * This method is invoked each time a message is logged by TSDuck.
         * @param severity Severity of the message.
         * @param message Message line.
         */
        @Override
        public void logMessageHandler(int severity, String message) {
            // Filter lines containing the JSON marker.
            final int pos = message.lastIndexOf(JSON_MARKER);
            if (pos >= 0) {
                // Pass the JSON text to the JSON processing routine.
                jsonHandler(message.substring(pos + JSON_MARKER.length()));
            }
            else {
                // This looks like a real log message.
                System.err.println(message);
            }
        }
    }

    /**
     * Main program.
     * @param args Command line arguments.
     */
    public static void main(String[] args) {

        // Transport stream file to analyze: command line argument, defaults to file.ts.
        final String fileName = args.length >= 1 ? args[0] : "file.ts";

        // Create a user-defined report to log multi-threaded messages.
        Logger rep = new Logger();

        // Create a TS processor using the report.
        TSProcessor tsp = new TSProcessor(rep);

        // Set the plugin chain.
        tsp.input = new String[] {"file", fileName};
        tsp.plugins = new String[][] {{"analyze", "--json-line=" + JSON_MARKER}};
        tsp.output = new String[] {"drop"};

        // Run the TS processing and wait until completion.
        tsp.start();
        tsp.waitForTermination();
        tsp.delete();

        // Terminate the asynchronous report.
        rep.terminate();
        rep.delete();
    }
}

File: sample/sample-python/sample-analyze-ts.py
#!/usr/bin/env python
#----------------------------------------------------------------------------
#
# TSDuck sample Python application : analyze a transport stream.
# A TSProcessor is used, using plugin "analyze" and JSON output.
# The JSON output is intercepted and parsed in a Python class.
# When found, the list of services in the transport stream is displayed.
# Any other information on the TS can be similarly extracted and used.
#
#----------------------------------------------------------------------------

import tsduck, json, sys

# This string is a user-defined marker to locate the JSON line in the log.
# It can be anything that is sufficiently weird to be unique in the logs.
json_marker = "@@_JSON_HERE_@@"

# This method processes the parsed JSON data from the TS analysis.
# Here, we just display the list of services.
def process_json(data):
    for srv in data['services']:
        print('Name: "%s", provider: "%s", bitrate: %d b/s' % (srv['name'], srv['provider'], srv['bitrate']))

# A Python class which handles TSDuck log messages.
class Logger(tsduck.AbstractAsyncReport):
    # This method is invoked each time a message is logged by TSDuck.
    def log(self, severity, message):
        # Filter, locate, extract and parse the JSON output from plugin "analyze".
        pos = message.find(json_marker)
        if pos >= 0:
            # This is a log line containing a JSON structure.
            process_json(json.loads(message[pos+len(json_marker):]))
        else:
            # This is a real log message, just display it.
            print(tsduck.Report.header(severity) + message, file=sys.stderr)

# Get input file from command line, default to file.ts.
input_file = sys.argv[1] if len(sys.argv) > 1 else 'file.ts'

# Main program: see details in "sample-tsp.py" or "sample-message-handling.py".
rep = Logger()
tsp = tsduck.TSProcessor(rep)

tsp.input = ['file', input_file]
tsp.plugins = [ ['analyze', '--json-line=' + json_marker] ]
tsp.output = ['drop']

tsp.start()
tsp.waitForTermination()
tsp.delete()

rep.terminate()
rep.delete()
